src/instruction/instructions/jump/bl.ts

```diff
@@ -0,0 +1,116 @@
+import { IMemory } from 'board/memory/interfaces'
+import { Register, Registers } from 'board/registers'
+import { ILabelOffsets } from 'instruction/interfaces'
+import { checkOptionCount, create, getBits, setBits } from 'instruction/opcode'
+import { Halfword, Word } from 'types/binary'
+import { BaseInstruction } from '../base'
+import { BinaryType } from 'types/binary/binaryType'
+
+/**
+ * Represents a 'BL' instruction.
+ */
+export class BlInstruction extends BaseInstruction {
+  public name: string = 'BL'
+  public pattern: string = '11110XXXXXXXXXXX'
+
+  public needsLabels: boolean = true
+  public opcodeLength: number = 2
+
+  private imm10Pattern: string = '000000XXXXXXXXXX'
+  private imm11Pattern: string = '00000XXXXXXXXXXX'
+  private patternSecondPart: string = '11X1XXXXXXXXXXXX'
+
+  public encodeInstruction(
+    options: string[],
+    labels?: ILabelOffsets
+  ): Halfword[] {
+    checkOptionCount(options, 1)
+    const offset = labels
+      ? Word.fromUnsignedInteger(labels[options[0]].value)
+      : Word.fromUnsignedInteger(0x0000)
+    const offsetBinaryString = offset.toBinaryString()
+    const sBit = offset.isBitSet(24)
+    const i1Bit = offset.isBitSet(23)
+    const i2Bit = offset.isBitSet(22)
+
+    let opcodeFirstPart: Halfword = create(this.pattern)
+    if (sBit) {
+      opcodeFirstPart = opcodeFirstPart.setBit(10)
+    }
+
+    opcodeFirstPart = setBits(
+      opcodeFirstPart,
+      this.imm10Pattern,
+      Halfword.fromUnsignedInteger(
+        parseInt(offsetBinaryString.substring(9, 20), 2)
+      )
+    )
+
+    let opcodeSecondPart: Halfword = create(this.patternSecondPart)
+    if (i1Bit == sBit) {
+      opcodeSecondPart = opcodeSecondPart.setBit(13)
+    }
+
+    if (i2Bit == sBit) {
+      opcodeSecondPart = opcodeSecondPart.setBit(11)
+    }
+
+    opcodeSecondPart = setBits(
+      opcodeSecondPart,
+      this.imm11Pattern,
+      Halfword.fromUnsignedInteger(
+        parseInt(offsetBinaryString.substring(19, 31), 2)
```

Comment:
![grafik](https://user-images.githubusercontent.com/56474048/165914743-ad2d03eb-9a04-4b5c-9077-7496f977f5fb.png)
for this offset -4 is provided so last 2 bits are set 0 in opcode only last bit (E) is set to 0

Reply:
my bad, sorry.

but still has some errors IMO
imm10 substring has length of 11 --> `offsetBinaryString.substring(9, 19)`
imm10 substring has length of 12 and overlaps with imm10 19th index from left --> `offsetBinaryString.substring(19, 30)`

also the s and i bits should be one higher since value is extended with zero on LSB

tests don't catch this since they are either negativ or irelevant on the certain bits

Reply:
Fixed. And added a test.